File: ghidra_scripts/ExportSelected.java
// Export selected functions and symbol XREFs for compact reverse-engineering review.
// @category LeCroy
//
// Usage from analyzeHeadless:
//   -postScript ExportSelected.java <output-dir> <target> [<target> ...]
//
// Targets may be addresses (for example 0x1619a or 1619a) or symbol names
// (for example KeSetEvent). Address targets export decompiled C plus compact
// incoming/outgoing function references. Symbol targets export references and
// the containing caller function names.

import java.io.File;
import java.io.PrintWriter;
import java.util.HashSet;
import java.util.Iterator;
import java.util.Set;

import ghidra.app.decompiler.DecompInterface;
import ghidra.app.decompiler.DecompileResults;
import ghidra.app.script.GhidraScript;
import ghidra.program.model.address.Address;
import ghidra.program.model.listing.Function;
import ghidra.program.model.listing.FunctionManager;
import ghidra.program.model.listing.Instruction;
import ghidra.program.model.listing.InstructionIterator;
import ghidra.program.model.symbol.Reference;
import ghidra.program.model.symbol.ReferenceIterator;
import ghidra.program.model.symbol.ReferenceManager;
import ghidra.program.model.symbol.Symbol;
import ghidra.program.model.symbol.SymbolIterator;
import ghidra.program.model.symbol.SymbolTable;

public class ExportSelected extends GhidraScript {
    private File outDir;
    private FunctionManager fm;
    private SymbolTable st;
    private ReferenceManager rm;
    private DecompInterface decomp;

    @Override
    public void run() throws Exception {
        String[] args = getScriptArgs();
        if (args.length < 2) {
            printerr("Usage: ExportSelected.java <output-dir> <target> [<target> ...]");
            return;
        }

        outDir = new File(args[0]);
        if (!outDir.exists() && !outDir.mkdirs()) {
            printerr("Failed to create output directory: " + outDir.getAbsolutePath());
            return;
        }

        fm = currentProgram.getFunctionManager();
        st = currentProgram.getSymbolTable();
        rm = currentProgram.getReferenceManager();

        decomp = new DecompInterface();
        decomp.openProgram(currentProgram);

        try {
            for (int i = 1; i < args.length; i++) {
                String target = args[i];
                Address addr = parseAddress(target);

                if (addr != null) {
                    Function f = fm.getFunctionAt(addr);
                    if (f == null) {
                        f = fm.getFunctionContaining(addr);
                    }

                    if (f == null) {
                        printerr("No function at/containing: " + target);
                    }
                    else {
                        writeFunction(f);
                    }
                }
                else {
                    writeSymbol(target);
                }
            }
        }
        finally {
            decomp.dispose();
        }

        println("ExportSelected.java: wrote results to " + outDir.getAbsolutePath());
    }

    private String sanitize(String s) {
        return s.replaceAll("[^A-Za-z0-9._-]", "_");
    }

    private Address parseAddress(String s) {
        String t = s.toLowerCase();
        try {
            long value;
            if (t.startsWith("0x")) {
                value = Long.parseUnsignedLong(t.substring(2), 16);
            }
            else if (t.matches("[0-9a-f]+")) {
                value = Long.parseUnsignedLong(t, 16);
            }
            else {
                return null;
            }
            return toAddr(value);
        }
        catch (Exception e) {
            return null;
        }
    }

    private Function functionAtOrContaining(Address addr) {
        Function f = fm.getFunctionAt(addr);
        if (f == null) {
            f = fm.getFunctionContaining(addr);
        }
        return f;
    }

    private void writeFunction(Function f) throws Exception {
        Address entry = f.getEntryPoint();
        String stem = entry + "_" + sanitize(f.getName());

        File cfile = new File(outDir, stem + ".c");
        try (PrintWriter pw = new PrintWriter(cfile, "UTF-8")) {
            DecompileResults res = decomp.decompileFunction(f, 60, monitor);
            if (res != null && res.decompileCompleted()) {
                pw.print(res.getDecompiledFunction().getC());
            }
            else {
                pw.println("/* decompilation failed */");
                if (res != null) {
                    pw.println("/* " + res.getErrorMessage() + " */");
                }
            }
        }

        File rfile = new File(outDir, stem + ".refs.txt");
        try (PrintWriter pw = new PrintWriter(rfile, "UTF-8")) {
            pw.println("FUNCTION " + entry + " " + f.getName());

            pw.println("INCOMING");
            Set<String> seenIncoming = new HashSet<>();
            ReferenceIterator refsTo = rm.getReferencesTo(entry);
            while (refsTo.hasNext()) {
                Reference r = refsTo.next();
                Function cf = functionAtOrContaining(r.getFromAddress());
                String line = r.getFromAddress() + " " +
                    (cf != null ? cf.getName() : "<no-function>");
                if (seenIncoming.add(line)) {
                    pw.println(line);
                }
            }

            pw.println("OUTGOING_FUNCTIONS");
            Set<String> seenOutgoing = new HashSet<>();
            InstructionIterator ins = currentProgram.getListing().getInstructions(f.getBody(), true);
            while (ins.hasNext()) {
                Instruction inst = ins.next();
                Reference[] refs = inst.getReferencesFrom();
                for (Reference r : refs) {
                    Function tf = fm.getFunctionAt(r.getToAddress());
                    if (tf != null) {
                        String line = tf.getEntryPoint() + " " + tf.getName();
                        if (seenOutgoing.add(line)) {
                            pw.println(line);
                        }
                    }
                }
            }
        }
    }

    private void writeSymbol(String name) throws Exception {
        SymbolIterator syms = st.getSymbols(name);
        int index = 0;
        boolean found = false;

        while (syms.hasNext()) {
            found = true;
            Symbol sym = syms.next();
            Address addr = sym.getAddress();

            File sfile = new File(outDir,
                "symbol_" + sanitize(name) + "_" + index + ".refs.txt");
            index++;

            try (PrintWriter pw = new PrintWriter(sfile, "UTF-8")) {
                pw.println("SYMBOL " + name + " " + addr);
                pw.println("REFERENCES");

                Set<String> seen = new HashSet<>();
                ReferenceIterator refsTo = rm.getReferencesTo(addr);
                while (refsTo.hasNext()) {
                    Reference r = refsTo.next();
                    Function cf = functionAtOrContaining(r.getFromAddress());
                    String line = r.getFromAddress() + " " +
                        (cf != null ? cf.getName() : "<no-function>");
                    if (seen.add(line)) {
                        pw.println(line);
                    }
                }
            }
        }

        if (!found) {
            printerr("Symbol not found: " + name);
        }
    }
}

File: ghidra_scripts/ExportSelected.py
#@category LeCroy
# Export selected functions and symbol XREFs for compact reverse-engineering review.
# Usage from analyzeHeadless:
#   -postScript ExportSelected.py <output-dir> <target> [<target> ...]
#
# Targets may be addresses (for example 0x1619a or 1619a) or symbol names
# (for example KeSetEvent). Address targets export decompiled C plus compact
# incoming/outgoing function references. Symbol targets export references and
# the containing caller function names.
#
# Written for Ghidra's Jython script environment.

from ghidra.app.decompiler import DecompInterface
from java.io import File, PrintWriter

args = getScriptArgs()
if len(args) < 2:
    printerr("Usage: ExportSelected.py <output-dir> <target> [<target> ...]")
    exit()

out_dir = File(args[0])
if not out_dir.exists():
    out_dir.mkdirs()

fm = currentProgram.getFunctionManager()
st = currentProgram.getSymbolTable()
rm = currentProgram.getReferenceManager()
listing = currentProgram.getListing()

decomp = DecompInterface()
decomp.openProgram(currentProgram)

def sanitize(s):
    return ''.join([c if c.isalnum() or c in '._-' else '_' for c in s])

def parse_addr(s):
    t = s.lower()
    try:
        if t.startswith("0x"):
            return toAddr(int(t[2:], 16))
        # Treat plain hex-looking targets as addresses.
        if all(c in "0123456789abcdef" for c in t):
            return toAddr(int(t, 16))
    except:
        pass
    return None

def function_name_at_or_containing(addr):
    f = fm.getFunctionAt(addr)
    if f is None:
        f = fm.getFunctionContaining(addr)
    return f

def write_function(f):
    entry = f.getEntryPoint()
    stem = "%s_%s" % (entry, sanitize(f.getName()))

    # Decompiled C
    cfile = File(out_dir, stem + ".c")
    pw = PrintWriter(cfile)
    try:
        res = decomp.decompileFunction(f, 60, monitor)
        if res is not None and res.decompileCompleted():
            pw.print(res.getDecompiledFunction().getC())
        else:
            pw.println("/* decompilation failed */")
            if res is not None:
                pw.println("/* %s */" % res.getErrorMessage())
    finally:
        pw.close()

    # Compact metadata / XREFs
    mfile = File(out_dir, stem + ".refs.txt")
    pw = PrintWriter(mfile)
    try:
        pw.println("FUNCTION %s %s" % (entry, f.getName()))

        pw.println("INCOMING")
        seen = set()
        it = rm.getReferencesTo(entry)
        while it.hasNext():
            r = it.next()
            cf = function_name_at_or_containing(r.getFromAddress())
            key = "%s %s" % (r.getFromAddress(), cf.getName() if cf else "<no-function>")
            if key not in seen:
                seen.add(key)
                pw.println(key)

        pw.println("OUTGOING_FUNCTIONS")
        seen = set()
        ins = listing.getInstructions(f.getBody(), True)
        while ins.hasNext():
            i = ins.next()
            for r in i.getReferencesFrom():
                tf = fm.getFunctionAt(r.getToAddress())
                if tf is not None:
                    key = "%s %s" % (tf.getEntryPoint(), tf.getName())
                    if key not in seen:
                        seen.add(key)
                        pw.println(key)
    finally:
        pw.close()

def write_symbol(name):
    syms = st.getSymbols(name)
    idx = 0
    found = False
    while syms.hasNext():
        found = True
        sym = syms.next()
        addr = sym.getAddress()
        sfile = File(out_dir, "symbol_%s_%d.refs.txt" % (sanitize(name), idx))
        idx += 1
        pw = PrintWriter(sfile)
        try:
            pw.println("SYMBOL %s %s" % (name, addr))
            pw.println("REFERENCES")
            seen = set()
            it = rm.getReferencesTo(addr)
            while it.hasNext():
                r = it.next()
                cf = function_name_at_or_containing(r.getFromAddress())
                key = "%s %s" % (r.getFromAddress(), cf.getName() if cf else "<no-function>")
                if key not in seen:
                    seen.add(key)
                    pw.println(key)
        finally:
            pw.close()
    if not found:
        printerr("Symbol not found: " + name)

for target in args[1:]:
    addr = parse_addr(target)
    if addr is not None:
        f = function_name_at_or_containing(addr)
        if f is None:
            printerr("No function at/containing: " + target)
        else:
            write_function(f)
    else:
        write_symbol(target)

decomp.dispose()
print("ExportSelected.py: wrote results to " + out_dir.getAbsolutePath())
